### PR TITLE
Fix git commit cache data curruption

### PR DESCRIPTION
**Pull request by yufeih**
Fixes https://github.com/dotnet/docfx/issues/3559 due to `FileShare` and `FileMode` are not atomic.

### CODE REVIEW

File: src/docfx/lib/git/GitCommitProvider.cs [hunk 1]
```diff
@@ -189,8 +189,11 @@ public Task SaveCache()
             {
                 using (var writer = new BinaryWriter(stream))
                 {
-                    writer.Write(_commitCache.Count);
-                    foreach (var (file, value) in _commitCache)
+                    // Create a snapshot of the commit cache to avoid writing corrupected data
+                    var commitCache = _commitCache.ToList();
```
Comment (OsmondJiang) [inline src/docfx/lib/git/GitCommitProvider.cs]: any reason here why ToList can resolve this problem?

Comment (yufeih) [reply src/docfx/lib/git/GitCommitProvider.cs]: after `writer.Write(_commitCache.Count)` and before `foreach (xxx in _commitCache)` is completed, there is a tiny chance that a new entry is added to `_commitCache`, causing the count on disk to not match the items count.

Comment (OsmondJiang) [reply src/docfx/lib/git/GitCommitProvider.cs]: so we should lock _commitCache right?

Comment (OsmondJiang) [reply src/docfx/lib/git/GitCommitProvider.cs]: I looked the code, didn't fond the tiny chance to change _commitCache during save unless we have some bugs, so we should take a look at the root cause

**Pull request merged**